### PR TITLE
migrate onedrive using prefix collection

### PR DESCRIPTION
**Pull request by ryanfkeepers**
#### Does this PR need a docs update or release note?

- [x] :clock1: Yes, but in a later PR

#### Type of change

- [x] :sunflower: Feature

#### Issue(s)

* #2825

#### Test Plan

- [x] :muscle: Manual
- [x] :zap: Unit test
- [x] :green_heart: E2E

### CODE REVIEW

File: src/internal/connector/onedrive/data_collections.go
Comment (ashmrtn) [inline src/internal/connector/onedrive/data_collections.go:128]: this is (probably?) going to need to work for sharepoint too

hmmm though now I see why this becomes logically complex :/
I'm a little hesitant to have ungated (by backup version or something) folder moves as I'm not sure if we'll ever want to tighten things down re: assumptions about moves (i.e. actually check there was something in the base moved when we thought there should be)

Eventually we may end up passing the previous backup version down to this function since we consume metadata, but we don't have the wiring for it right now to help gate this. Ideally, we should just design some interface that allows passing this (and other) info down to stuff. We could maybe place it on the RestoreCollection, but it feels a tad out of place there since all collections should have the same backup version

It would make me uncomfortable, but we can go ahead with this. We definitely need a ticket to track using this only when we actually need to migrate something though

File: src/internal/kopia/upload.go
Comment (ashmrtn) [inline src/internal/kopia/upload.go:993]: nit: add a comment that this handles migration of the prefix directories. That way the comment below that explains that it assumes no prefix directory changes makes sense (idk how I missed that earlier 🙃)

File: src/pkg/path/path.go
Comment (ashmrtn) [inline src/pkg/path/path.go:287]: why require a builder instance at all? Seems like you could just do `path.ToServicePrefix` and save some keystrokes

File: src/internal/kopia/path_encoder.go
Comment (ashmrtn) [inline src/internal/kopia/path_encoder.go:29]: this should be returned as a real error. Or this should be in a test file and it should be clear it's only for test functions

**Review by ashmrtn — COMMENTED** [review]
looks good with a few nits. Some discussion needed around how to gate this if possible though

File: src/pkg/path/path.go
Comment (ryanfkeepers) [reply src/pkg/path/path.go:287]: Goooood question.

File: src/internal/connector/onedrive/data_collections.go
Comment (ryanfkeepers) [reply src/internal/connector/onedrive/data_collections.go:128]: > this is (probably?) going to need to work for sharepoint too

As a general statement?  Or did you have a particular case in mind?

> Eventually we may end up passing the previous backup version down to this function since we consume metadata

I was thinking of doing that for this change already.  Pushing unwarranted migrations sounds like a fantastic footgun.  Wanted to hold off for this PR and re-evaluate today/later, since it's a two-way door.

> passing this (and other) info down to stuff

I waffled on that, myself.  There's an ownership tug-of-war between backup/repo versioning/design and collection sourcing.  If we're going to keep with the collection-based migration, the clearest owner would be the package that produces those collections, and what we'd want to pass down is an indicator of which migration to perform.  Operations is agnostic to the granular details of collection data, and passing in an interface from collections/data_collections is a bit moot.

> We definitely need a ticket to track using this only when we actually need to migrate something though

I'll either follow up today or make a ticket for it by eod.

File: src/internal/connector/graph/collections.go
Comment (ashmrtn) [inline src/internal/connector/graph/collections.go:67]: hmmm looking at this with fresh eyes this needs a comment and a fixup

fixup: it's not going to work for collections denoting deleted folders because they have a nil `FullPath` so I think it will NPE

comment: this will not guarantee getting all collections, only those that aren't deleted (see above). It may be less confusing to only track collections related to the prefix directories?

File: src/internal/connector/onedrive/data_collections.go
Comment (ashmrtn) [reply src/internal/connector/onedrive/data_collections.go:128]: > >this is (probably?) going to need to work for sharepoint too
>
>As a general statement? Or did you have a particular case in mind?

There's probably SharePoint backups out there that will need migrating since I believe we've pointed at least a few folks to that feature so we likely need the ability to handle that

>Wanted to hold off for this PR and re-evaluate today/later, since it's a two-way door.

are you suggesting merge this and fixup later? If we do that I think we should have a solid timeline on when "later" is. As you said, it's a fantastic footgun

File: src/internal/kopia/upload_test.go
Comment (ashmrtn) [inline src/internal/kopia/upload_test.go:201]: nit: would be nice to say something about expecting "at least" since this is a subset check

File: src/pkg/path/path_test.go
Comment (ashmrtn) [inline src/pkg/path/path_test.go:752]: can you also make a test for calling `Folders()` on a `ServicePrefix` path? I'm not sure if that will throw an index out of bounds error right now since it's kind of escaping the original assumptions of Path

**Review by ashmrtn — COMMENTED** [review]
LGTM except for a few minor things. Need to know more about the plan re: gating this with a backup version or something

File: src/internal/connector/onedrive/data_collections.go
Comment (ryanfkeepers) [reply src/internal/connector/onedrive/data_collections.go:128]: > There's probably SharePoint backups out there that will need migrating since I believe we've pointed at least a few folks to that feature so we likely need the ability to handle that

I think the plan for sharepoint is not to include migrations, since it's barely been released at all.

> when "later" is.

already patched: https://github.com/alcionai/corso/pull/3137

File: src/pkg/path/path_test.go
Comment (ryanfkeepers) [reply src/pkg/path/path_test.go:752]: good idea

File: src/internal/connector/onedrive/data_collections.go
Comment (ryanfkeepers) [reply src/internal/connector/onedrive/data_collections.go:131]: @ashmrtn 3137 has been squashed onto this PR.

**Comment by ryanfkeepers** [comment]
Side note: the details.go file is feeling a little cluttered at this point.  Would be nice to see a follow-up that splits the various structures/handlers into their own files.

**Comment by ashmrtn** [comment]
just to put this here as it was surfaced offline: this needs a little bit of coordination prior to merging to make sure everyone is alright with it going in

Removing the code that actually triggers the migration (i.e. version bump plus code in the repo package I think) would allow us to merge all the supporting logic while holding off on the actual migration

File: src/internal/connector/onedrive/data_collections.go
Comment (ashmrtn) [reply src/internal/connector/onedrive/data_collections.go:131]: >I think the plan for sharepoint is not to include migrations, since it's barely been released at all.

It may be better to just plan for SP migrations and handle it if needed. That way there's no surprises down the line

File: src/internal/version/backup.go
Comment (ashmrtn) [inline src/internal/version/backup.go:44]: this needs updated

**Review by ashmrtn — CHANGES_REQUESTED** [review]
putting on a block on this since the version needs updated

**Comment by ryanfkeepers** [comment]
> putting on a block on this since the version needs updated

@ashmrtn please see: https://github.com/alcionai/corso/pull/3187

File: src/internal/connector/onedrive/data_collections.go
Comment (ryanfkeepers) [reply src/internal/connector/onedrive/data_collections.go:131]: > It may be better to just plan for SP migrations and handle it if needed.

Oh, I have it all planned out.  The problem with SP, what we don't have with OD, is that SP would migrate from a value that isn't passed to the caller onto one that is.  The migration itself would be an easy lift on top of this.  It's the additional plumbing specifically to surface data for that migration which I'd like to avoid.

**Comment by ryanfkeepers** [comment]
@ashmrtn all PRs are in, this should be ready for a final OK.

File: src/internal/connector/onedrive/collection.go
Comment (ashmrtn) [inline src/internal/connector/onedrive/collection.go:196]: since we switched to using the graph package's prefix collection I don't think this function should be required anymore?

File: src/internal/connector/onedrive/data_collections.go
Comment (ashmrtn) [inline src/internal/connector/onedrive/data_collections.go:140]: nit: could use `version.NoBackup`?

File: src/internal/connector/onedrive/data_collections_test.go
Comment (ashmrtn) [inline src/internal/connector/onedrive/data_collections_test.go:81]: this seems like it could be simplified. It seems like it should be either 1 or 0 collections so the `LessOrEqual` check should be removed and we can directly do a `Len` check instead of checking for nil

File: src/internal/connector/onedrive/data_collections_test.go
Comment (ashmrtn) [inline src/internal/connector/onedrive/data_collections_test.go:83]: The below could maybe be simplified to make `migr` instances in a slice `[]migr` and then do a `Contains` check at the bottom to ensure the migration collection is in there

File: src/internal/connector/sharepoint/data_collections.go
Comment (ashmrtn) [inline src/internal/connector/sharepoint/data_collections.go:120]: I would feel better if we passed in `collections` here as expected instead of nil
```suggestion
			collections,
```

File: src/internal/kopia/upload_test.go
Comment (ashmrtn) [inline src/internal/kopia/upload_test.go:198]:
```suggestion
		require.Lenf(t, entries, len(dirs), "expected exactly %+v\ngot %+v", dd, de)
```

File: src/internal/kopia/upload_test.go
Comment (ashmrtn) [inline src/internal/kopia/upload_test.go:2708]: these lines don't apply to this test
```suggestion
```

File: src/internal/kopia/upload_test.go
Comment (ashmrtn) [inline src/internal/kopia/upload_test.go:2717]: these need updated

File: src/internal/operations/backup.go
Comment (ashmrtn) [inline src/internal/operations/backup.go:43]: can you make a comment saying this is mostly expected to be used for tests?

File: src/internal/version/backup.go
Comment (ashmrtn) [inline src/internal/version/backup.go:48]: I'm a little on the fence about what the value of this should be. Right now, my understanding is setting it to `Backup + 1` will cause the OD migration code to always create a migration collection but the paths will be the same as selectors is wired to have the UPN for both. If we switch this to `NoBackup` though then we'll have no way to test the migration code :/

I'm not sure if there's a good path forward, do you have any thoughts on this?

File: src/internal/common/idname/idname.go
Comment (ryanfkeepers) [reply src/internal/common/idname/idname.go:27]: huh?  Yeah!  what gives?  Where did that commit go?

File: src/internal/common/idname/idname.go
Comment (ryanfkeepers) [reply src/internal/common/idname/idname.go:47]: I swear I did... now I'm worried.

File: src/internal/connector/onedrive/data_collections.go
Comment (ryanfkeepers) [reply src/internal/connector/onedrive/data_collections.go:140]: made a helper func for it.

File: src/internal/connector/onedrive/data_collections_test.go
Comment (ryanfkeepers) [reply src/internal/connector/onedrive/data_collections_test.go:81]: I'm apparently too scarred by integration tests to remember I don't need to worry about it in unit testing. (:

File: src/internal/connector/sharepoint/data_collections.go
Comment (ryanfkeepers) [reply src/internal/connector/sharepoint/data_collections.go:120]: oh good catch, thanks.

File: src/internal/version/backup.go
Comment (ryanfkeepers) [reply src/internal/version/backup.go:48]: @ashmrtn see: https://github.com/alcionai/corso/pull/3195

File: src/internal/common/idname/idname.go
Comment (ryanfkeepers) [reply src/internal/common/idname/idname.go:47]: 😬 yeah.  here's the [merged, unexported version](https://github.com/alcionai/corso/pull/3158/commits/6ce3577c6edcc33c1083f99a1c77fa0efd9525c1).  And here's the [version merged to this branch](https://github.com/alcionai/corso/pull/3158/files#diff-8fd9848f6b38951a8e702aa0e07e5a4e0b3d532a9685f736b900e33edc5713dbR35).

/panic

Comment (ryanfkeepers) [reply src/internal/common/idname/idname.go:47]: https://github.com/alcionai/corso/pull/3196

File: src/internal/version/backup.go
Comment (ryanfkeepers) [reply src/internal/version/backup.go:48]: 3195 is merged

File: src/internal/common/idname/idname.go
Comment (ryanfkeepers) [reply src/internal/common/idname/idname.go]: 3196 is merged.

I am once again asking for a final review on this PR.

**Review by ashmrtn — APPROVED** [review]

**Pull request merged**
